EmptyPage와 Layout, Pagination 컴포넌트 작업

hanana1253 opened the pull request:
## PR Type

What kind of change does this PR introduce?

<!-- 'x'를 이용하여 이 PR에 적용되는 항목을 확인해 주세요. -->

- [ ] 버그를 수정했어요.
- [x] 새로운 기능을 추가했어요.
- [ ] 코드 스타일 업데이트를 했어요(포맷팅, 지역변수)
- [ ] 리팩토링을 했어요 (기능적인 변화 없이, api 변경 없이)
- [ ] 환경설정을 변경했어요.
- [ ] 문서 내용을 변경했어요.
- [ ] 기타 사항을 설명해 주세요.

## Related Issues

solved #34 
solved #37 
<!--#을 눌러 이슈와 연결해 주세요-->

## What does this PR do?

<!--무엇을 하셨나요?-->

- [x] PageNotFound, 결과 없음 등에 쓰일 수 있는 EmptyPage 컴포넌트 작성 및 스타일링
- [x] Layout 틀만 잡아두었습니다 (header는 작업 필요)
- [x] Pagination 컴포넌트를 만들었습니다. 

## Other information

<!--참고한 자료, 추가적인 사항, 기타 의견-->
- [Link 태그 안에 react-router-dom처럼 a를 자동으로 생성해주는 게 아니라 a태그를 넣어주어야 하는 이유](https://uchanlee.dev/nextjs/Why-using-a-tag-in-nextjs-Link/)

hanana1253 committed:
feat: Header 컴포넌트 틀 만들기 및  스토리 작성

TODO:
- Header에 맞는 로고와 검색인풋, 버튼 등 넣어야합니다.
- 스타일링 해야합니다.

hanana1253 committed:
feat: Layout 컴포넌트 만들고 스토리 작성

- 모든 페이지에 렌더링 될 Header을 가지고 있는 Layout 컴포넌트를 만들었습니다.
- scroll to top 버튼 만들면 추가하기 (기존 코드에서는 header에 포함되어있었으나 별도로 두는 게 좋을듯)

hanana1253 committed:
chore: 컴포넌트 index 파일 경로 추가, Layout 추가

hanana1253 committed:
fix: Loading 컴포넌트를 위한 돔 요소 div 추가

- dom 요소를 추가했습니다.

hanana1253 committed:
feat: EmptyPage 컴포넌트를 만들었습니다

- 404 페이지 및 검색결과/마이레시피 결과 없을 때 보여줄 EmptyPage입니다.

hanana1253 committed:
fix: 404 페이지를 EmptyPage 컴포넌트로 구현

- 스타일링이 들어간 Empty Page 컴포넌트로 만들었습니다.
:

hanana1253 committed:
Merge branch 'feature/no-result' into develop

hustle-dev commented on src/components/EmptyPage/EmptyPage.types.ts:
```diff
@@ -0,0 +1,3 @@
+export interface EmptyPageProps {
+  children: JSX.Element | JSX.Element[];
```
이렇게 작성하게되면 string을 안받겠다는 의지인 것 같아요

참고사이트

- https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/basic_type_example#useful-react-prop-type-examples

hustle-dev commented on src/components/EmptyPage/EmptyPage.stories.tsx:
```diff
@@ -0,0 +1,24 @@
+  args: {
+    headline: 'Nothing Saved',
+    description: 'Save your recipe!',
+  },
```
이 부분은 더이상 필요없을것 같습니다

hustle-dev commented on src/components/Layout/Layout.tsx:
```diff
@@ -0,0 +1,13 @@
+      <Header />
+      <main>{children}</main>
+      <div id="loading-start" aria-live="assertive" />
+      <div id="loading-end" aria-live="assertive" />
```
loading 부분 주석처리하셨으면 이 부분 나중에 제가 다시 고쳐보도록하겠습니다.

없애두 될거같아요

hustle-dev commented on src/pages/404.tsx:
```diff
@@ -1,7 +1,14 @@
+import EmptyPage from 'components/EmptyPage/EmptyPage';
```
re-export에 추가해서 가져올 수 있을것 같아요

hustle-dev commented on src/pages/404.tsx:
```diff
@@ -1,7 +1,14 @@
+import Link from 'next/link';
 import { NextPage } from 'next';
 
 const NotFound: NextPage = () => {
```
404 next.js config를 활용하여 redirect하는부분도 있으면 좋을거같아요

hanana1253 replied:
특정 url을 redirect하는 건 next.config.js에서 쉽게 할 수 있는 것 같은데 없는 페이지에 대해서 일괄적으로 404로 redirect하는 것은 조사해봐야 할 것 같아요.

hanana1253 committed:
fix: Type정의할 때  children JSX.Element => ReactNode로 변경

hanana1253 committed:
fix: EmptyPage 컴포넌트 리뷰 내용에 따른 수정

- 404에서 불러올 때 경로 수정
- component/index.ts에서 re-export 구문 추가
- 스토리북에서 decorator 추가 및 변경된 타입구조 적용
- styling 코드에서 theme으로 primaryOrange 컬러 적용

hanana1253 committed:
fix: Loading을 위한 div 요소 삭제

hanana1253 committed:
fix: Import Path 잘못 표기한 story 파일 수정

- default export를 수정하면서 경로 변경되어 story파일을 수정했습니다.

hanana1253 committed:
feat: Pagination 컴포넌트 작성

- className 프롭은 emotion을 사용하므로 더이상 필요없게 되어 삭제
- 앞 페이지 목록으로 넘어가는 ellipsis 버튼이 딱히 사용성이 있게 느껴지지 않아 삭제

hanana1253 committed:
Merge branch 'feature/pagination' into develop

hanana1253 committed:
fix: 404 page Link 태그 안에 a 태그 추가

- 공식문서 및 사용법을 보니 Link 태그와 a 태그를 함께 쓰네요

hanana1253 commented on src/pages/404.tsx:
```diff
@@ -1,7 +1,16 @@
+    <EmptyPage>
+      <h2>Page Not Found</h2>
+      <Link href="/">
+        <a>Go to main page?</a>
```
정민님이 지적해주신 것처럼 Link태그 안에 요소가 들어오는 경우엔 a 태그로 꼭 감싸줘야 하더라구요.
children이 text로만 넘어오면 <a> 태그로 감싸주긴 하지만 일관적인 사용을 위해 a 태그 추가했습니다.
이유는 아래 블로그에 잘 설명되어있는데 블로그 쓰신 분 열정이 대단해서 참고해보시면 좋을 것 같아요.
https://uchanlee.dev/nextjs/Why-using-a-tag-in-nextjs-Link/

hustle-dev replied:
넵넵 한번 확인해보겠습니다 👍

hustle-dev replied:
ㄷㄷ 블로그 글 쓰신분 확인하려고 내부 코드까지 뜯어보다니 대단하시네여..

hustle-dev commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,52 @@
+export const Pagination = ({ limit, currentPage, onClick: handleClick, totalResults }: PaginationProps) => {
+  const lastPageNum = Math.ceil(totalResults / limit);
+  const pageStartNum = Math.max(currentPage - 2, 1);
+  const pageEndNum = Math.min(currentPage + 2, lastPageNum);
```
이 부분을 분리해서 커스텀훅을 만들어 연산이 많이 필요한 경우 useMemo를 사용해도 될것 같아요 🙂

hanana1253 replied:
좋은 아이디어네요 바로 반영하겠습니다.
